waspc/data/Generator/templates/react-app/src/actions/core.js

```diff
@@ -1,10 +1,19 @@
 import { callOperation } from '../operations'
-import { invalidateQueriesUsing } from '../operations/resources'
+import {
+  registerActionInProgress,
+  registerActionDone,
+} from '../operations/resources'
 
 export function createAction(actionRoute, entitiesUsed) {
-  return async function action(args) {
+  async function internalAction(args, optimisticallyUpdatedCacheKeys = []) {
+    registerActionInProgress(optimisticallyUpdatedCacheKeys)
     const actionResult = await callOperation(actionRoute, args)
-    await invalidateQueriesUsing(entitiesUsed)
+    await registerActionDone(entitiesUsed, optimisticallyUpdatedCacheKeys)
     return actionResult
   }
+
+  const action = (args) => internalAction(args)
+  action.internal = internalAction
```

Comment:
An internal action that can take extra arguments (currently only optimistically updated cache keys).

Since we don't want to expose this to our users, I should change the string `internal` to a `Symbol`.